File: client/src/utils/statusConstants.js
/**
 * Status Constants for Frontend
 * This file contains all status IDs used in the frontend to match the backend StatusConstants
 */

export const STATUS_CONSTANTS = {
  // General Status IDs
  ACTIVE: 1,
  INACTIVE: 2,

  // Task Status IDs
  TASK_PENDING: 1,
  TASK_IN_PROGRESS: 2,
  TASK_COMPLETED: 12,
  TASK_REJECTED: 6,
  TASK_SUBMITTED_ON_TIME: 10,
  TASK_SUBMITTED_LATE: 11,
  TASK_COMPLETED_LATE: 13,

  // Report Status IDs
  REPORT_SUBMITTED: 1,
  REPORT_CONFIRMED: 5,

  // Grade Status IDs
  GRADE_ACTIVE: 1,

  // Account Status IDs
  ACCOUNT_ACTIVE: 1,
  ACCOUNT_INACTIVE: 2,

  // Role IDs (for reference)
  ROLE_SUPERVISOR: 3,
  ROLE_TEACHER: 4,
  ROLE_ENGINEER: 5,
};

/**
 * Helper functions for status management
 */
export const StatusHelpers = {
  /**
   * Gets the appropriate task status based on submission timing
   * @param {boolean} isLate - Whether the submission is late
   * @returns {number} Status ID for submission timing
   */
  getTaskSubmissionStatus: (isLate) => {
    return isLate ? STATUS_CONSTANTS.TASK_SUBMITTED_LATE : STATUS_CONSTANTS.TASK_SUBMITTED_ON_TIME;
  },

  /**
   * Gets the appropriate task completion status based on timing
   * @param {boolean} isCompletedLate - Whether the completion is late
   * @returns {number} Status ID for completion timing
   */
  getTaskCompletionStatus: (isCompletedLate) => {
    return isCompletedLate ? STATUS_CONSTANTS.TASK_COMPLETED_LATE : STATUS_CONSTANTS.TASK_COMPLETED;
  },

  /**
   * Gets the default report status
   * @returns {number} Default report status ID
   */
  getDefaultReportStatus: () => {
    return STATUS_CONSTANTS.REPORT_SUBMITTED;
  },

  /**
   * Gets status text based on status ID
   * @param {number} statusId - The status ID
   * @param {string} deadline - The deadline string (optional)
   * @param {boolean} isPendingTask - Whether this is a pending task
   * @param {boolean} isLateFromServer - Whether the server has determined this is late (preferred)
   * @returns {string} Status text
   */
  getStatusText: (statusId, deadline = null, isPendingTask = false, isLateFromServer = null) => {
    console.log("StatusHelpers.getStatusText - statusId:", statusId, "deadline:", deadline, "isPendingTask:", isPendingTask, "isLateFromServer:", isLateFromServer);
    
    // Handle pending tasks without submissions
    if (isPendingTask) {
      if (deadline && isDeadlinePassed(deadline, isLateFromServer)) {
        return "Deadline Passed";
      }
      return "Pending";
    }
    
    switch (statusId) {
      case STATUS_CONSTANTS.TASK_COMPLETED:
        return deadline && isDeadlinePassed(deadline, isLateFromServer) ? "Completed Late" : "Completed";
      case STATUS_CONSTANTS.TASK_REJECTED:
        console.log("StatusHelpers.getStatusText - TASK_REJECTED case matched, returning 'Rejected'");
        return "Rejected";
      case STATUS_CONSTANTS.TASK_COMPLETED_LATE:
        return "Completed Late";
      case STATUS_CONSTANTS.TASK_SUBMITTED_LATE:
        return "Submitted Late";
      case STATUS_CONSTANTS.TASK_SUBMITTED_ON_TIME:
        return deadline && isDeadlinePassed(deadline, isLateFromServer) ? "Submitted Late" : "Submitted On Time";
      case STATUS_CONSTANTS.TASK_PENDING:
        return "Pending";
      default:
        return "Pending";
    }
  },

  /**
   * Gets status color based on status ID
   * @param {number} statusId - The status ID
   * @param {string} deadline - The deadline string (optional)
   * @param {boolean} isPendingTask - Whether this is a pending task
   * @param {boolean} isLateFromServer - Whether the server has determined this is late (preferred)
   * @returns {string} Status color
   */
  getStatusColor: (statusId, deadline = null, isPendingTask = false, isLateFromServer = null) => {
    // Handle pending tasks without submissions
    if (isPendingTask) {
      if (deadline && isDeadlinePassed(deadline, isLateFromServer)) {
        return "#e53e3e"; // Red for deadline passed
      }
      return "#d69e2e"; // Yellow for pending
    }
    
    switch (statusId) {
      case STATUS_CONSTANTS.TASK_COMPLETED:
        return deadline && isDeadlinePassed(deadline, isLateFromServer) ? "#d69e2e" : "#38a169"; // Yellow for late, Green for on-time
      case STATUS_CONSTANTS.TASK_REJECTED:
        return "#e53e3e"; // Red
      case STATUS_CONSTANTS.TASK_SUBMITTED_ON_TIME:
        return deadline && isDeadlinePassed(deadline, isLateFromServer) ? "#e53e3e" : "#3182ce"; // Red for late, Blue for on-time
      case STATUS_CONSTANTS.TASK_SUBMITTED_LATE:
        return "#e53e3e"; // Red for Submitted Late
      case STATUS_CONSTANTS.TASK_COMPLETED_LATE:
        return "#d69e2e"; // Yellow for Completed Late
      case STATUS_CONSTANTS.TASK_PENDING:
        return "#d69e2e"; // Yellow
      default:
        return "#d69e2e"; // Yellow
    }
  }
};

/**
 * Helper function to check if deadline has passed
 * WARNING: This function uses client-side time which can be manipulated by users
 * For security, always rely on server-side deadline validation
 * @param {string} deadlineString - The deadline string
 * @param {boolean} isLateFromServer - Whether the server has determined this is late (preferred)
 * @returns {boolean} Whether the deadline has passed
 */
const isDeadlinePassed = (deadlineString, isLateFromServer = null) => {
  // If server has already determined the status, use that (more secure)
  if (isLateFromServer !== null) {
    return isLateFromServer;
  }
  
  // Fallback to client-side check (less secure, can be manipulated)
  if (!deadlineString) return false;
  
  try {
    const deadline = new Date(deadlineString);
    const now = new Date();
    return now > deadline;
  } catch (error) {
    console.error("Error checking deadline:", error);
    return false;
  }
};

File: server/Elsewedy-Capstone_System/Constants/StatusConstants.cs
using System;

namespace Elsewedy_Capstone_System.Constants
{
    /// <summary>
    /// Contains all status IDs used throughout the system for easy management and modification
    /// </summary>
    public static class StatusConstants
    {
        #region General Status IDs
        
        /// <summary>
        /// Active/Enabled status
        /// </summary>
        public const long Active = 1L;
        
        /// <summary>
        /// Inactive/Disabled status
        /// </summary>
        public const long Inactive = 2L;
        
        #endregion

        #region Task Status IDs
        
        /// <summary>
        /// Task is pending/not started
        /// </summary>
        public const long TaskPending = 1L;
        
        /// <summary>
        /// Task is in progress
        /// </summary>
        public const long TaskInProgress = 2L;
        
        /// <summary>
        /// Task is completed successfully
        /// </summary>
        public const long TaskCompleted = 12L;
        
        /// <summary>
        /// Task submission is rejected
        /// </summary>
        public const long TaskRejected = 6L;
        
        /// <summary>
        /// Task submitted on time
        /// </summary>
        public const long TaskSubmittedOnTime = 10L;
        
        /// <summary>
        /// Task submitted late
        /// </summary>
        public const long TaskSubmittedLate = 11L;
        
        /// <summary>
        /// Task completed late
        /// </summary>
        public const long TaskCompletedLate = 13L;
        
        #endregion

        #region Report Status IDs
        
        /// <summary>
        /// Report is submitted (default status)
        /// </summary>
        public const long ReportSubmitted = 1L;
        
        /// <summary>
        /// Report is under review
        /// </summary>
        public const long ReportUnderReview = 2L;
        
        /// <summary>
        /// Report is approved
        /// </summary>
        public const long ReportApproved = 3L;
        
        /// <summary>
        /// Report is rejected
        /// </summary>
        public const long ReportRejected = 4L;
        
        #endregion

        #region Grade Status IDs
        
        /// <summary>
        /// Grade is active
        /// </summary>
        public const long GradeActive = 1L;
        
        #endregion

        #region Account Status IDs
        
        /// <summary>
        /// Account is active
        /// </summary>
        public const long AccountActive = 1L;
        
        /// <summary>
        /// Account is inactive
        /// </summary>
        public const long AccountInactive = 2L;
        
        #endregion

        #region Role IDs (for reference)
        
        /// <summary>
        /// Supervisor role ID
        /// </summary>
        public const long RoleSupervisor = 3L;
        
        /// <summary>
        /// Teacher role ID
        /// </summary>
        public const long RoleTeacher = 4L;
        
        /// <summary>
        /// Engineer role ID
        /// </summary>
        public const long RoleEngineer = 5L;
        
        #endregion

        #region Helper Methods
        
        /// <summary>
        /// Gets the appropriate task status based on submission timing
        /// </summary>
        /// <param name="isLate">Whether the submission is late</param>
        /// <returns>Status ID for submission timing</returns>
        public static long GetTaskSubmissionStatus(bool isLate)
        {
            return isLate ? TaskSubmittedLate : TaskSubmittedOnTime;
        }
        
        /// <summary>
        /// Gets the appropriate task completion status based on timing
        /// </summary>
        /// <param name="isCompletedLate">Whether the completion is late</param>
        /// <returns>Status ID for completion timing</returns>
        public static long GetTaskCompletionStatus(bool isCompletedLate)
        {
            return isCompletedLate ? TaskCompletedLate : TaskCompleted;
        }
        
        /// <summary>
        /// Gets the default report status
        /// </summary>
        /// <returns>Default report status ID</returns>
        public static long GetDefaultReportStatus()
        {
            return ReportSubmitted;
        }
        
        #endregion
    }
}
